### PR TITLE
Introducing IdentityDataStoreService

### PR DESCRIPTION
**Pull request by ZiyamSanthosh**
### Proposed changes in this pull request

- With this PR, a new OSGI service named IdentityDataStoreService will be introduced which will contain method related to storing identity claims to IDN_IDENTITY_USER_DATA table.

### Related issue
- https://github.com/wso2/product-is/issues/15974

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            identityDataStore = (UserIdentityDataStore) clazz.newInstance();
+        }
+    }
+
+    @Override
+    public boolean storeInIdentityDataStore(String userName, UserStoreManager userStoreManager, String operationType,
+                                            Map<String, String> claims) throws UserStoreException {
+
+        // Check if the user is locked.
+        if (PRE_SET_USER_CLAIM_VALUES.equals(operationType)) {
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java:55]: What is the purpose of this "account lock" check if the rest of the method body does not impact?

Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java:55]: This is to denote that the account will be locked by storing these identity claims. This was the previous behaviour as well.

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                if (!StringUtils.equalsIgnoreCase(operationType, PRE_USER_ADD_CLAIM_VALUES)) {
+                    // We avoid loading claims for pre user add operations.
+                    userIdentityClaim = identityDataStore.load(userName, userStoreManager);
+                }
+
+                if (userIdentityClaim == null) {
+                    userIdentityClaim = new UserIdentityClaim(userName);
+                }
+
+                Iterator<Map.Entry<String, String>> it = claims.entrySet().iterator();
```
[the remark this answers is not in the document]
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java]: Fixed with https://github.com/wso2-extensions/identity-governance/pull/717/commits/acfcbd236e0d9ed5b65130101fb233b70ee5585c

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ */
+public class IdentityDataStoreServiceImpl implements IdentityDataStoreService {
+
+    private UserIdentityDataStore identityDataStore;
+
+    private static final String PRE_SET_USER_CLAIM_VALUES = "PreSetUserClaimValues";
+    private static final String PRE_USER_ADD_CLAIM_VALUES = "PreAddUserClaimValues";
+    private static final String STORE_IDENTITY_CLAIMS = "StoreIdentityClaims";
+    private static final String IDENTITY_DATA_STORE_TYPE = "IdentityDataStore.DataStoreType";
+    private static final String DEFAULT_JDBC_IDENTITY_DATA_STORE = "org.wso2.carbon.identity.governance.store" +
```
[the remark this answers is not in the document]
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java]: Fixed with https://github.com/wso2-extensions/identity-governance/pull/717/commits/acfcbd236e0d9ed5b65130101fb233b70ee5585c

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                    userIdentityClaim = new UserIdentityClaim(userName);
+                }
+
+                Iterator<Map.Entry<String, String>> claimsIterator = claims.entrySet().iterator();
+
+                while (claimsIterator.hasNext()) {
+                    Map.Entry<String, String> claim = claimsIterator.next();
+                    String key = claim.getKey();
+                    String value = claim.getValue();
+                    if (key.contains(UserCoreConstants.ClaimTypeURIs.IDENTITY_CLAIM_URI_PREFIX)) {
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java:90]: this is prone to NPEs

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
+package org.wso2.carbon.identity.governance.service;
```
[the remark this answers is not in the document]
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java:19]: Fixed with https://github.com/wso2-extensions/identity-governance/pull/717/commits/21ecba47eebc28b5b1490d86f27de56db0eb78cf

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        }
+    }
+
+    @Override
+    public boolean storeInIdentityDataStore(String userName, UserStoreManager userStoreManager, String operationType,
+                                            Map<String, String> claims) throws UserStoreException {
+
+        // Check if the user is locked.
+        if (PRE_SET_USER_CLAIM_VALUES.equals(operationType)) {
+            boolean accountLocked = Boolean.parseBoolean(claims.get(UserIdentityDataStore.ACCOUNT_LOCK));
```
[the remark this answers is not in the document]
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java]: Fixed with https://github.com/wso2-extensions/identity-governance/pull/717/commits/21ecba47eebc28b5b1490d86f27de56db0eb78cf

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreService.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
+package org.wso2.carbon.identity.governance.service;
```
[the remark this answers is not in the document]
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreService.java:19]: Fixed with https://github.com/wso2-extensions/identity-governance/pull/717/commits/21ecba47eebc28b5b1490d86f27de56db0eb78cf

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java (hunk 4 above)
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java:111]: Fixed with https://github.com/wso2-extensions/identity-governance/pull/717/commits/21ecba47eebc28b5b1490d86f27de56db0eb78cf

**Review by chamathns — APPROVED** [review]

**Review by indeewari — APPROVED** [review]

**Pull request merged**